Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV1.cpp

```diff
@@ -0,0 +1,680 @@
+            return;
+
+        auto* priv = WPE_IM_CONTEXT_WAYLAND_V1(global->current)->priv;
+
+        uint32_t wpe_modifiers = 0;
+        if (modifiers & priv->modifiers.shiftMask)
+            wpe_modifiers |= WPE_MODIFIER_KEYBOARD_SHIFT;
+        if (modifiers & priv->modifiers.altMask)
+            wpe_modifiers |= WPE_MODIFIER_KEYBOARD_ALT;
+        if (modifiers & priv->modifiers.controlMask)
+            wpe_modifiers |= WPE_MODIFIER_KEYBOARD_CONTROL;
+
+        auto* event = wpe_event_keyboard_new(state ? WPE_EVENT_KEYBOARD_KEY_DOWN : WPE_EVENT_KEYBOARD_KEY_UP, priv->view,
+            WPE_INPUT_SOURCE_KEYBOARD, time, static_cast<WPEModifiers>(wpe_modifiers), 0, sym);
+        wpe_view_event(priv->view, event);
+        wpe_event_unref(event);
+    },
+    // language
+    [](void* /*data*/, struct zwp_text_input_v1* /*text_input*/, uint32_t /*serial*/, const char* /*language*/)
+    {
+        // NOOP
+    },
+    // text_direction
+    [](void* /*data*/, struct zwp_text_input_v1* /*text_input*/, uint32_t /*serial*/, uint32_t /*direction*/)
+    {
+        // NOOP
+    }
+};
+
+static void wpeIMContextWaylandV1GlobalFree(gpointer data)
+{
+    auto* global = static_cast<TextInputV1Global*>(data);
+    g_free(global);
+}
+
+static TextInputV1Global* textInputV1GetGlobalByDisplay(WPEDisplayWayland *display)
+{
+    auto* global = static_cast<TextInputV1Global*>(g_object_get_data(G_OBJECT(display), "text-input-v1-global"));
+    if (global != nullptr)
+        return global;
+
+    global = g_new0(TextInputV1Global, 1);
+    global->textInput = wpeDisplayWaylandGetTextInputV1(display);
+
+    if (global->textInput)
+        zwp_text_input_v1_add_listener(global->textInput, &textInputListenerV1, global);
+
+    g_object_set_data_full(G_OBJECT(display),
+        "text-input-v1-global",
+        global,
+        wpeIMContextWaylandV1GlobalFree);
+    return global;
+}
+
+static void wpeIMContextWaylandV1PurposeOrHintsChanged(WPEInputMethodContext *context)
+{
+    textInputV1SetContentType(WPE_IM_CONTEXT_WAYLAND_V1(context));
+    textInputV1CommitState(WPE_IM_CONTEXT_WAYLAND_V1(context));
+}
+
+static void wpeIMContextWaylandV1Constructed(GObject* object)
+{
+    G_OBJECT_CLASS(wpe_im_context_wayland_v1_parent_class)->constructed(object);
+
+    auto* context = WPE_INPUT_METHOD_CONTEXT(object);
+    auto* wpeContext = WPE_IM_CONTEXT_WAYLAND_V1(context);
+
+    g_signal_connect_swapped(context, "notify::input-purpose", G_CALLBACK(wpeIMContextWaylandV1PurposeOrHintsChanged), wpeContext);
+    g_signal_connect_swapped(context, "notify::input-hints", G_CALLBACK(wpeIMContextWaylandV1PurposeOrHintsChanged), wpeContext);
+}
+
+static void wpeIMContextWaylandV1Finalize(GObject* object)
+{
+    auto* self = WPE_IM_CONTEXT_WAYLAND_V1(object);
+
+    wpe_im_context_focus_out(WPE_INPUT_METHOD_CONTEXT(self));
+
+    g_clear_object(&self->priv->view);
+    g_free(self->priv->preedit.text);
+    g_free(self->priv->surrounding.text);
```

Comment:
Use GUniquePtr or CString

Reply:
There's also GList pointer for underlines which I just noticed that is not handled by finalize. I think that's something that should be freed explicitly even after changing text char* pointers to GUniquePtr?

Reply:
Fixed